### PR TITLE
add support for generate_release_notes

### PR DESCRIPTION
**Pull request by mball-agathos**
Issue: #2794

### CODE REVIEW

**Comment by mball-agathos** [comment]
Hi @EnricoMi !  If you have a moment at some point, this PR should be ready to ship, with 100% incremental test coverage and practical testing from my related projects.  Thanks!

File: .github/test_release_notes.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# This file is needed for generating release notes for the Repository::testGenerateReleaseNotesWithAllArguments test
```
Comment (EnricoMi) [inline .github/test_release_notes.yml:1]: Can we use the existing release notes generated here?
https://github.com/PyGithub/PyGithub/blob/main/.github/release-drafter.yml

File: tests/ReplayData/Repository.testGenerateReleaseNotes.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+https
+POST
+api.github.com
+None
+/repos/jacquev6/PyGithub/releases/generate-notes
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotes.txt:5]: Are these test files hand crafted?
```suggestion
/repos/PyGithub/PyGithub/releases/generate-notes
```

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: tests/ReplayData/Repository.testGenerateReleaseNotes.txt (hunk 2 above)
Comment (mball-agathos) [reply tests/ReplayData/Repository.testGenerateReleaseNotes.txt:5]: These files were generated with and then hand-edited.  I generated them using an OAuth token and then edited them to conform to the password-generated format.  Based on the previous testing setup, these were configured to use `jacquev6` in the test setup, so I edited the files to also use `jacquev6` so that the unit tests would pass.

This testing framework is novel to me, so there's a good chance I've done something wrong!

Here's my command for generating the ReplayData:

`pytest tests/Repository.py::Repository::testGenerateReleaseNotesWithAllArguments --record --auth_with_token`

I just now attempted to update `jacquev6` to `PyGithub` in the two newly test files, and I get this resulting error in the Repository test:

```
========================================================================= short test summary info ==========================================================================
FAILED Repository.py::Repository::testGenerateReleaseNotes - requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
FAILED Repository.py::Repository::testGenerateReleaseNotesWithAllArguments - requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
================================================================ 2 failed, 937 passed, 2 warnings in 30.15s ================================================================
```

I did a quick search for `jacquev6`, and found that this appears 100s of times in the test files, so it appears that these are pretty much expected to use `jacquev6` instead of the more appropriate `PyGitHub`.

So, in short, changing `jacquev6` to `PyGitHub` would be a much larger scope PR than this one (although it would probably be a really good cleanup!)

File: .github/test_release_notes.yml (hunk 1 above)
Comment (mball-agathos) [reply .github/test_release_notes.yml:1]: Note that this `release-drafter.yml` config is based on this 3rd-party GitHub action: https://github.com/release-drafter/release-drafter .

Conversely, the test in this PR is based on GitHub's own built-in release notes yaml defined here: https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes

Since the functionality introduced in this PR is specifically testing GitHub's implementation, I don't think we want to use a third-party version!

All that said, this technically probably doesn't need to go into the `.github` folder (although admittedly I haven't confirmed this).  I can move this into another folder, like something under the `tests` folder if you have thoughts for a better place!  I could also just delete this file, but keep in mind that to regenerate the test replay data, it's necessary for this file to be checked into the `main` branch before running the test (because the GitHub API itself requires this template to already be in the default branch)

**Comment by mball-agathos** [comment]
Thank you @EnricoMi for your review!  Based on my responses, please let me know if there are any changes that you'd like to see!

**Comment by mball-agathos** [comment]
Hi @EnricoMi !  Are there any other changes you'd like to see in this PR?  I'm hoping to avoid bit rot! :)

File: github/Repository.py [hunk 3]
```diff
@@ -1342,6 +1343,46 @@ def create_git_release(
...
+            "tag_name": tag_name,
+        }
+        if is_defined(previous_tag_name):
+            post_parameters["previous_tag_name"] = str(previous_tag_name)
+
+        if is_defined(target_commitish):
+            post_parameters["target_commitish"] = str(target_commitish)
+
+        if is_defined(configuration_file_path):
+            post_parameters["configuration_file_path"] = str(configuration_file_path)
```
Comment (EnricoMi) [inline github/Repository.py:1378]:
```suggestion
        post_parameters = NotSet.remove_unset_items({
            "tag_name": tag_name,
            "previous_tag_name": previous_tag_name,
            "target_commitish": target_commitish,
            "configuration_file_path": configuration_file_path,
        })
```

File: github/Repository.py [hunk 4]
```diff
@@ -1342,6 +1343,46 @@ def create_git_release(
...
+        assert isinstance(target_commitish, str) or is_optional(target_commitish, str), target_commitish
+        assert isinstance(configuration_file_path, str) or is_optional(
+            configuration_file_path, str
+        ), configuration_file_path
+
+        post_parameters = {
+            "tag_name": tag_name,
+        }
+        if is_defined(previous_tag_name):
+            post_parameters["previous_tag_name"] = str(previous_tag_name)
```
Comment (EnricoMi) [inline github/Repository.py:1372]: Is there any reason for the `str(...)`?

File: tests/ReplayData/Repository.testGenerateReleaseNotes.txt (hunk 2 above)
Comment (EnricoMi) [reply tests/ReplayData/Repository.testGenerateReleaseNotes.txt:5]: Right. In the meantime, main branch moved `Repository` tests from `jacquev6` to `PyGithub`, so you can easily do this replacement in your test files.

File: .github/test_release_notes.yml (hunk 1 above)
Comment (EnricoMi) [reply .github/test_release_notes.yml:1]: Yes, please move this file to `tests/`.

File: .github/test_release_notes.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+# See https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes
+
```
Comment (EnricoMi) [inline .github/test_release_notes.yml:3]: Please note which test uses this `yaml` file.

**Review by EnricoMi — CHANGES_REQUESTED** [review]
Minor comments / changes required.

**Comment by mball-agathos** [comment]
Thanks you @EnricoMi for your comments!  I should be able to address these later this week!

File: .github/test_release_notes.yml (hunk 5 above)
Comment (mball-agathos) [reply .github/test_release_notes.yml]: Done

File: .github/test_release_notes.yml (hunk 1 above)
Comment (mball-agathos) [reply .github/test_release_notes.yml]: Done

File: github/Repository.py (hunk 4 above)
Comment (mball-agathos) [reply github/Repository.py]: I'm not sure if this is needed, but I've accepted the code suggestion that makes it unnecessary.

**Review by mball-agathos — COMMENTED** [review]
Thank you for the review!  I don't think I've got the unit tests quite working, but I ran out of time this weekend, so I'm just uploading what I was able to change so far.  The new testing framework changes over the last 6 months have shifted just enough that I'll have to wrap my head around this project again to get them working.

**Comment by Dolphindalt** [comment]
I have an important use case for this function and really need it. :pray: Anything I can do to help things progress?

**Comment by mball-agathos** [comment]
Hi @Dolphindalt 

Thank you for offering your help!  What I'm currently blocked on is that my test no longer works correctly, and I'm having a really hard time debugging it.  I think you can help by regenerating the ReplayData for the tests (at least, this is my most optimistic path towards getting them working again).  The ReplayData was generated using instructions from https://github.com/PyGithub/PyGithub/blob/main/CONTRIBUTING.md#automated-tests (specifically, running `pytest` with the `--record` flag), and then doing a little bit of search-and-replay from my GitHub user name to `jacquev6` (or whatever the latest tests are using now).

Thanks!

**Comment by Dolphindalt** [comment]
There are some potential issues in the `Framework.py` class that seem to be causing the `Repository.testGenerateReleaseNotesWithAllArguments.txt` test to fail. I've resolved these problems but am stuck on authenticating. The API call does not allow for personal access tokens and with 2 factor authentication, using a login and password is not possible. How does one use a JWT to authenticate to use the API?

File: tests/ReplayData/Repository.testGenerateReleaseNotes.txt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+https
+POST
+api.github.com
+None
+/repos/PyGithub/PyGithub/releases/generate-notes
+{'Authorization': 'Basic login_and_password_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotes.txt]:
```suggestion
{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```

File: tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+https
+POST
+api.github.com
+None
+/repos/PyGithub/PyGithub/releases
+{'Authorization': 'Basic login_and_password_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt]:
```suggestion
{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```

File: tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+https
+POST
+api.github.com
+None
+/repos/PyGithub/PyGithub/releases/generate-notes
+{'Authorization': 'Basic login_and_password_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt]:
```suggestion
{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
```

File: tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt [hunk 9]
```diff
@@ -0,0 +1,21 @@
+https
+POST
+api.github.com
+None
+/repos/PyGithub/PyGithub/releases
+{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
+{"tag_name": "vX.Y.Z-by-PyGithub-acctest-previous", "draft": false, "prerelease": false, "generate_release_notes": false, "name": "vX.Y.Z: PyGithub acctest", "body": "This release is created by PyGithub"}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt]:
```suggestion
{"tag_name": "vX.Y.Z-by-PyGithub-acctest-previous", "draft": false, "prerelease": false, "generate_release_notes": false, "name": "vX.Y.Z: PyGithub acctest", "body": "This release is created by PyGithub", "make_latest": "true"}
```

File: tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+{"name":"vX.Y.Z-by-PyGithub-acctest","body":"<!-- Release notes generated using configuration in .github/test_release_notes.yml at main -->\n\n\n\n**Full Changelog**: https://github.com/PyGithub/PyGithub/compare/vX.Y.Z-by-PyGithub-acctest-previous...vX.Y.Z-by-PyGithub-acctest"}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt]:
```suggestion
{"name":"vX.Y.Z-by-PyGithub-acctest","body":"<!-- Release notes generated using configuration in tests/test_release_notes.yml at main -->\n\n\n\n**Full Changelog**: https://github.com/PyGithub/PyGithub/compare/vX.Y.Z-by-PyGithub-acctest-previous...vX.Y.Z-by-PyGithub-acctest"}
```

File: tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+https
+POST
+api.github.com
+None
+/repos/PyGithub/PyGithub/releases/generate-notes
+{'Authorization': 'token private_token_removed', 'User-Agent': 'PyGithub/Python', 'Content-Type': 'application/json'}
+{"tag_name": "vX.Y.Z-by-PyGithub-acctest", "previous_tag_name": "vX.Y.Z-by-PyGithub-acctest-previous", "target_commitish": "main", "configuration_file_path": ".github/test_release_notes.yml"}
```
Comment (EnricoMi) [inline tests/ReplayData/Repository.testGenerateReleaseNotesWithAllArguments.txt]:
```suggestion
{"tag_name": "vX.Y.Z-by-PyGithub-acctest", "previous_tag_name": "vX.Y.Z-by-PyGithub-acctest-previous", "target_commitish": "main", "configuration_file_path": "tests/test_release_notes.yml"}
```

File: tests/Repository.py [hunk 12]
```diff
@@ -537,6 +537,34 @@ def testCreateGitReleaseWithAllArguments(self):
...
+        )
+        notes = self.repo.generate_release_notes(
+            tag_name="vX.Y.Z-by-PyGithub-acctest",
+            previous_tag_name="vX.Y.Z-by-PyGithub-acctest-previous",
+            target_commitish="main",
+            configuration_file_path="tests/test_release_notes.yml",
+        )
+        self.assertEqual(notes.name, None)
+        self.assertIn(
+            "Release notes generated using configuration in .github/test_release_notes.yml at main", notes.body
```
Comment (EnricoMi) [inline tests/Repository.py]:
```suggestion
            "Release notes generated using configuration in tests/test_release_notes.yml at main", notes.body
```

File: github/GeneratedReleaseNotes.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+
+# New file named github/GeneratedReleaseNotes.py
+class GeneratedReleaseNotes(CompletableGithubObject):
+    """
+    This class represents the release notes generated by the release/generate-notes REST API endpoint.
+
+    The reference can be found here:
+    https://docs.github.com/en/rest/releases/releases#generate-release-notes-content-for-a-release
+
```
Comment (EnricoMi) [inline github/GeneratedReleaseNotes.py:18]:
```suggestion

    The OpenAPI schema can be found at
    - /components/schemas/release-notes-content

```

**Comment by EnricoMi** [comment]
Can you run `pre-commit run --all-files` to fix the lint errors please?

**Comment by mball-agathos** [comment]
@EnricoMi I really appreciate your help in getting this fixed!  I just haven't had the time to get my environment setup again to properly debug the new testing issue.

I'm running `pre-commit run --all-files` now, but am encountering an issue where both `docformatter` and `black` are introducing conflicting formatting changes.  I think I can work through it, but it's also possible that we can't run both tools on these particular changes.

**Comment by mball-agathos** [comment]
For my local setup, running `pre-commit run --all-files` caused a lot of collateral reformatting, and an endless fight between black and docformatter.  However, in looking at the lint failure, I saw that just two Python files needed reformatting, so I ran black just on those two.

@EnricoMi At this point, it will need your approval again to run the CI/CD.  I'll let you press the "Update branch" button to retrigger these tests. :)

**Comment by EnricoMi** [comment]
Please fix the order of attributes and methods as indicated here:
https://github.com/PyGithub/PyGithub/actions/runs/17202921596/job/48797288811?pr=3022#step:5:53-80

**Comment by mball-agathos** [comment]
I've fixed the attribute ordering, but I'm having a little trouble understanding the fix for "Check verbs": https://github.com/PyGithub/PyGithub/actions/runs/17202921596/job/48797288802?pr=3022

```
Head verb checking:
  Indexing 169 Python files
  Indexed 221 classes
  Indexed 356 paths
  Indexed 239 schemas
Base verb checking:
  Indexing 168 Python files
  Indexed 220 classes
  Indexed 355 paths
  Indexed 238 schemas

Difference:
1,4c1,4
< Indexing 168 Python files
< Indexed 220 classes
< Indexed 355 paths
< Indexed 238 schemas
---
> Indexing 169 Python files
> Indexed 221 classes
> Indexed 356 paths
> Indexed 239 schemas
```

**Comment by mball-agathos** [comment]
From digging a little more into the "Check verbs" failure, it looks to me like this may have been a side-effect of the PR not being fully up-to-date with `main`, or failures in unrelated files.  But as reference, here are the fixes that Cursor suggested to get this action working again.  Note in particular that scripts/openapi.py seems to have syntax errors in `main` currently:

```diff
diff --git a/github/Organization.py b/github/Organization.py
index 1a40272a..a8c20e86 100644
--- a/github/Organization.py
+++ b/github/Organization.py
@@ -1329,7 +1329,8 @@ class Organization(CompletableGithubObject):
 
     def get_copilot(self) -> Copilot:
         """
-        :calls: Various Copilot-related endpoints for this organization :rtype: :class:`github.Copilot.Copilot`
+        :calls: `GET /orgs/{org}/copilot/billing/seats <https://docs.github.com/en/rest/copilot/copilot-business>`_
+        :rtype: :class:`github.Copilot.Copilot`
         """
         return github.Copilot.Copilot(self._requester, self.login)
 
diff --git a/github/openapi.index.json b/github/openapi.index.json
index 8c2260fa..0201649c 100644
--- a/github/openapi.index.json
+++ b/github/openapi.index.json
@@ -8,6 +8,7 @@
     "GithubIntegration.get_user_installation": "GET",
     "GithubIntegration.get_app_installation": "GET",
     "Organization.create_fork": "POST",
+    "Organization.get_copilot": "GET",
     "PullRequest.create_comment": "POST",
     "PullRequest.get_comment": "GET",
     "PullRequest.get_comments": "GET",
diff --git a/scripts/openapi.py b/scripts/openapi.py
index 0c8b70b8..8410b240 100644
--- a/scripts/openapi.py
+++ b/scripts/openapi.py
@@ -2386,7 +2386,7 @@ class OpenApi:
                                         .get("methods")
                                     ]
                                     if self.verbose:
-                                        print(f"    - {verb} {candidate_path} implemented by {", ".join(methods)}")
+                                        print(f"    - {verb} {candidate_path} implemented by {', '.join(methods)}")
                                     continue
 
                                 suggested_methods = self.suggest_method_names(verb, path, candidate_path, spec)
@@ -2433,7 +2433,7 @@ class OpenApi:
 
             spec_fingerprints = {fingerprint(path): path for path in spec_paths}
             for path in sorted(list(unspec_paths)):
-                print(f"- {path}: {spec_fingerprints.get(fingerprint(path), " ")}")
+                print(f"- {path}: {spec_fingerprints.get(fingerprint(path), ' ')}")
         print()
 
     def suggest_method_names(self, verb: str, prefix_path: str, path: str, spec: dict[str, Any]) -> list[str]:

```

**Comment by EnricoMi** [comment]
The "Check verbs" differences are side-effects of your PR. Nothing to worry about.

Thanks for the syntax errors, I'll look into them.

File: github/GeneratedReleaseNotes.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    The OpenAPI schema can be found at
+    - /components/schemas/release-notes-content
```
Comment (EnricoMi) [inline github/GeneratedReleaseNotes.py:20]:
```suggestion

        - /components/schemas/release-notes-content
```

File: github/GeneratedReleaseNotes.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
+"""
+Representation of the generated release notes returned by the release/generate-notes REST API endpoint.
+"""
+
```
Comment (EnricoMi) [inline github/GeneratedReleaseNotes.py]:
```suggestion
from __future__ import annotations

```

File: github/GeneratedReleaseNotes.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+from github.GithubObject import (
+    Attribute,
+    CompletableGithubObject,
+    NotSet,
+)
+
+
+# New file named github/GeneratedReleaseNotes.py
+class GeneratedReleaseNotes(CompletableGithubObject):
```
Comment (EnricoMi) [inline github/GeneratedReleaseNotes.py]:
```suggestion
class GeneratedReleaseNotes(NonCompletableGithubObject):
```

File: github/GeneratedReleaseNotes.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Any, Dict
+
+from github.GithubObject import (
+    Attribute,
+    CompletableGithubObject,
+    NotSet,
+)
+
+
+# New file named github/GeneratedReleaseNotes.py
```
Comment (EnricoMi) [inline github/GeneratedReleaseNotes.py]:
```suggestion
```

File: github/Repository.py [hunk 18]
```diff
@@ -1549,6 +1550,43 @@ def create_git_release(
...
+                "target_commitish": target_commitish,
+                "configuration_file_path": configuration_file_path,
+            }
+        )
+
+        headers, data = self._requester.requestJsonAndCheck(
+            "POST", f"{self.url}/releases/generate-notes", input=post_parameters
+        )
+
+        return GeneratedReleaseNotes(self._requester, headers, data, completed=True)
```
Comment (EnricoMi) [inline github/Repository.py]:
```suggestion
        return GeneratedReleaseNotes(self._requester, headers, data)
```

**Review by EnricoMi — APPROVED** [review]
LGTM! Thanks for your work!

**Comment by mball-agathos** [comment]
Thanks @EnricoMi for helping me get this across the finish line! :)
